### PR TITLE
DEIMOS missing slits and alignment boxes

### PR DESCRIPTION
**Pull request by deborape**
This is an upgrade to the slitmask matching procedure that I have implemented earlier.
The changes are a decent number, but they are mostly re-arranging stuff around.

Main changes:

- Removing traces that are not expected from slitmask design.
- Adding slits that are expected from the slitmask design but not found in the tracing process.
- Flaging as “BOXSLIT” slits that are expected to be alignment boxes from slitmask design.
- Dealing with overlapping alignment boxes.
- Added an astropy Table to the SlitTraceSet datamodel with slitmask design and objects info. (SlitTraceSet v1.1.0 -> v1.1.1)
- Cleaning (commenting) some unused slitmask matching ParSets
- Added unit test `test_maskdesign_matching::test_add_missing_slits`

Run `./pypeit_test develop -i keck_deimos -t 2` and passed. 

```
Reducing data from keck_deimos for the following setups:
    600ZD_M_6500
    600ZD_tilted
    1200G_M_7750
    830G_LVM_8400
    830G_M_8100_26
    830G_M_8500
    830G_L_8100

Running tests in 2 parallel processes
 0 passed/ 0 failed/ 0 skipped STARTED keck_deimos/830G_M_8500 pypeit (without masters)
 0 passed/ 0 failed/ 0 skipped STARTED keck_deimos/600ZD_M_6500 pypeit (without masters)
 1 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/600ZD_M_6500 pypeit (without masters)
 1 passed/ 0 failed/ 0 skipped STARTED keck_deimos/1200G_M_7750 pypeit (without masters)
 2 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/830G_M_8500 pypeit (without masters)
 2 passed/ 0 failed/ 0 skipped STARTED keck_deimos/600ZD_tilted pypeit (without masters)
 3 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/600ZD_tilted pypeit (without masters)
 3 passed/ 0 failed/ 0 skipped STARTED keck_deimos/830G_L_8100 pypeit (without masters)
 4 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/1200G_M_7750 pypeit (without masters)
 4 passed/ 0 failed/ 0 skipped STARTED keck_deimos/830G_LVM_8400 pypeit (without masters)
 5 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/830G_LVM_8400 pypeit (without masters)
 5 passed/ 0 failed/ 0 skipped STARTED keck_deimos/830G_M_8100_26 pypeit (without masters)
 6 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/830G_L_8100 pypeit (without masters)
 7 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/830G_M_8100_26 pypeit (without masters)

--- PYPEIT DEVELOPMENT SUITE PASSED 7/7 TESTS (Masters ignored) ---
```

I will run the entire Dev Suite and generate a new Cooked (since I edited the SlitTraceSet datamodel). After that I will be able to run all the unit tests.

Run `./update_docs` but got 3 warnings that I am not able to understand (also they don’t seem related to this PR). See below.

```
WARNING: autodoc: failed to import module 'sciencecube' from module 'pypeit.images'; the following exception was raised:
No module named 'shapely'
/Users/dpelliccia/GitHub_Pypeit/PypeIt/doc/dev/development.rst:490: WARNING: Unknown target name: "conda_forge".
```

### CODE REVIEW

File: pypeit/edgetrace.py
Comment (profxj) [inline pypeit/edgetrace.py:4217]: Please add an inline comment or two.

File: pypeit/edgetrace.py
Comment (profxj) [inline pypeit/edgetrace.py:4109]: Seems we might need a `while` or `for` loop here instead of nested if/then
in case there are 4!   :) 

But I'm ok with this for now.

File: pypeit/edgetrace.py
Comment (profxj) [inline pypeit/edgetrace.py:4315]: As this effectively defines the DataModel for this Table, 
kindly add details to the doc string on what is each column.

File: pypeit/edgetrace.py
Comment (profxj) [inline pypeit/edgetrace.py:4362]: Same comment as above.  Please define precisely the Table columns.

I should ask the same in one of the `.rst` files.  Thanks.

File: pypeit/par/pypeitpar.py
Comment (profxj) [inline pypeit/par/pypeitpar.py:2788]: If @kbwestfall agrees, remove these.

File: pypeit/slittrace.py
Comment (profxj) [inline pypeit/slittrace.py:261]: @kbwestfall can comment on whether there is a more elegant way to do this.
Might not be.

**Review by profxj — CHANGES_REQUESTED** [review]
Most minor things from me. 

Main one is to document the datamodel of the maskdesign Table
saved in SlitEdgeTrace.  i.e.  what are the columns (names and
data types)?

Thanks!

File: pypeit/edgetrace.py
Comment (profxj) [reply pypeit/edgetrace.py:4362]: Note:  I am adding `OBJNAME` in the `deimos_radec` branch.

**Comment by deborape** [comment]
I run the Dev Suite and I got only one failure:

```
--- PYPEIT DEVELOPMENT SUITE FAILED 1/89 TESTS (Masters ignored) ---
Failed tests:
    keck_nires/NIRES pypeit_ql_mos
Skipped tests:
Testing Started at 2020-10-27T18:53:20.844378
Testing Completed at 2020-10-28T18:57:09.692972
Total Time: 1 day, 0:03:48.848594
```

```
Error Messages:

End of Log:
  File "/Users/dpelliccia/GitHub_Pypeit/PypeIt/pypeit/pypmsgs.py", line 267, in error
    raise PypeItError(msg)
pypeit.pypmsgs.PypeItError: Current version of SlitTraceSet object in code (v1.1.1) does not match version used to write your HDU(s)!
```

The reason is that `pypeit_ql_mos` uses Master frames from the folder `QL_MASTERS/NIRES_MASTERS/` and there `MasterSlits_A_1_01.fits.gz`  uses still an old version of SlitTraceSet (that I have updated in this PR). How should I go about it?

I also generated a new Cooked and run the unit tests and passed:
`=================== 253 passed in 806.85s (0:13:26) ===============`

**Comment by profxj** [comment]
You may ignore the `NIRES_MASTERS` test.  It is not relevant to
this PR.

File: pypeit/edgetrace.py
Comment (deborape) [reply pypeit/edgetrace.py:4109]: Sorry, I wasn't able to find another way to code this part.

File: pypeit/edgetrace.py
Comment (deborape) [reply pypeit/edgetrace.py:4217]: Done.

File: pypeit/edgetrace.py
Comment (deborape) [reply pypeit/edgetrace.py:4315]: Done.

File: pypeit/edgetrace.py
Comment (deborape) [reply pypeit/edgetrace.py:4362]: Done the doc string. I will update this to include `OBJNAME` once I move to the `deimos_radec` branch.

I am not sure which `.rst` file I should update. Can you point it out to me?

File: pypeit/edgetrace.py
Comment (deborape) [reply pypeit/edgetrace.py:4936]: I changed this to be able to generate a `design_table` and `object_table`, even if some `maskdef_id` is `-99` (this should be very rare). In this case the tables will not include the traces for which we could not assign a `maskdef_id`. However, `_maskdef_id` in `SlitTraceSet ` will still display the -99 for the traces with no assigned id.

File: pypeit/edgetrace.py
Comment (profxj) [reply pypeit/edgetrace.py:4108]: No worries.

File: pypeit/edgetrace.py
Comment (profxj) [reply pypeit/edgetrace.py:4387]: If the items are specific to DEIMOS then use `deimos.rst`  
If not, then I thknk `slit_tracing.rst` would be good.
Thanks

File: pypeit/edgetrace.py
Comment (kbwestfall) [reply pypeit/edgetrace.py:4108]: As long as you're both okay with it, I think it's okay if we punt on this for now.  But...

- Note you can swap values in python without an intermediate variable.  Something like, `x, y = y, x`.

- @profxj , I thought we decided that you wanted the edges of the overlapping slits to instead be the same so that the flexure correction doesn't have gaps in the spatial slit illumination pattern that you use in the cross-correlation.  I don't think this does this, right?

- We probably do need something general that can deal with more than just 3 overlapping slits.  Can we instead check this based on the optical model and then throw an error if that's the case?  That way, we don't need to write the more general code now, but we'll know if the code ever runs across a dataset where this is an issue.

File: pypeit/edgetrace.py
Comment (kbwestfall) [inline pypeit/edgetrace.py:4986]: Ah, here it is.  So, can we just use `edge_msk` instead of also having `align_slit`?

**Review by kbwestfall — APPROVED** [review]
Looks good.  I didn't dig too deep, but I had a few recommendations.  I'm approving now, so that @profxj can give the final approval before merging.

File: pypeit/edgetrace.py
Comment (deborape) [reply pypeit/edgetrace.py:491]: Thank you for the suggestion. I did remove the attribute `align_slit` and used the 'BOXSLIT' flag instead.

I also added these new attributes to the doc string

File: pypeit/edgetrace.py
Comment (deborape) [reply pypeit/edgetrace.py:3959]: I did revert to `edge_msk`

File: pypeit/edgetrace.py
Comment (deborape) [reply pypeit/edgetrace.py:4108]: Ok. I did change this part of the code. Instead of switching the overlapping traces, what we do now is to move the left and right edges to be adjacent. I tested with the case of 3 overlapping slits and works well there too.

File: pypeit/edgetrace.py
Comment (deborape) [reply pypeit/edgetrace.py:4986]: Done!

File: pypeit/par/pypeitpar.py
Comment (deborape) [reply pypeit/par/pypeitpar.py:2797]: Done, thanks!

File: pypeit/edgetrace.py
Comment (deborape) [reply pypeit/edgetrace.py:4400]: Ok! I decided to add a doc for MasterSlits that was missing, and updates some other docs. Please let me know if you think everything is alright.

**Review by profxj — APPROVED** [review]
merge when ready!

**Comment by profxj** [comment]
merging

**Pull request merged**
